vim-patch:8.2.1535: it is not possible to specify cell widths of characters

delphinus opened the pull request:
Problem:    It is not possible to specify cell widths of characters.
Solution:   Add setcellwidths().
https://github.com/vim/vim/commit/08aac3c6192f0103cb87e280270a32b50e653be1

delphinus committed:
vim-patch:8.2.1535: it is not possible to specify cell widths of characters

Problem:    It is not possible to specify cell widths of characters.
Solution:   Add setcellwidths().
https://github.com/vim/vim/commit/08aac3c6192f0103cb87e280270a32b50e653be1

delphinus commented:
TODO: Later I'll commit for `v8.2.1537: memory acccess error when using setcellwidths()` here.

delphinus committed:
Fix for linting

bfredl commented on src/nvim/eval/funcs.c:
```diff
@@ -8175,6 +8176,116 @@ static void f_setbufvar(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+  return li1->li_tv.vval.v_number - li2->li_tv.vval.v_number;
+}
+
+// "setcellwidths()" function
```
three slashes

delphinus replied:
Fixed on f942ce6

bfredl commented on src/nvim/mbyte.c:
```diff
@@ -458,6 +458,55 @@ static bool intable(const struct interval *table, size_t n_items, int c)
+cw_interval_T *cw_table = NULL;
+size_t cw_table_size = 0;
+
+// Return 1 or 2 when "c" is in the cellwidth table.
```
three slashes and use `@return`

delphinus replied:
Fixed on f37505b

delphinus committed:
Merge branch 'master' into vim-8.2.1535

bfredl commented on src/nvim/mbyte.c:
```diff
@@ -472,12 +521,31 @@ static bool intable(const struct interval *table, size_t n_items, int c)
+    if (n != 0) {
+      return n;
+    }
+#ifdef USE_WCHAR_FUNCTIONS
+    //
+    // Assume the library function wcwidth() works better than our own
+    // stuff.  It should return 1 for ambiguous width chars!
+    //
+    n = wcwidth(c);
+
+    if (n < 0) {
+      return 6;                 // unprintable, displays <xxxx>
+    }
+    if (n > 1) {
+      return n;
+    }
+#else
```
this should be deleted

delphinus replied:
Fixed on 7a2620506

delphinus committed:
fix: Use 3 slashes to describe the function name

https://github.com/neovim/neovim/pull/13883/files#r585428210

delphinus committed:
fix: Add more detail for the function

https://github.com/neovim/neovim/pull/13883/files#r585428707

delphinus committed:
fix: Remove already removed #ifdef

https://github.com/neovim/neovim/pull/18618

delphinus committed:
fix: Do not overwrite existing names

delphinus committed:
ci: Use semsg() to avoid errors

delphinus committed:
vim-patch:8.2.3545: setcellwidths() may make 'listchars' or 'fillchars' invalid

Problem:    setcellwidths() may make 'listchars' or 'fillchars' invalid.
Solution:   Check the value and give an error. (closes vim/vim#9024)
https://github.com/vim/vim/commit/94358a1e6e640ca5ebeb295efdddd4e92b700673

delphinus committed:
fix: Do not use func the same as vim does

delphinus committed:
vim-patch:8.2.1537: memory acccess error when using setcellwidths()

Problem:    Memory acccess error when using setcellwidths().
Solution:   Use array and pointers correctly.
https://github.com/vim/vim/commit/b06a6d59d12dbd67d55b3c46f6e5547e9103c931

delphinus committed:
fix: indent

delphinus committed:
fix: Deal with local value of 'fillchars'

ref https://github.com/vim/vim/commit/96ba25ac01279f73c0ecb5d4aa4ff37aa359e5eb

delphinus committed:
fix: Use valid types to avoid warning

delphinus committed:
fix: reference to vars & funcs

delphinus committed:
fix: typo

zeertzjq commented on src/nvim/eval/funcs.c:
```diff
@@ -8132,6 +8132,156 @@ static void set_position(typval_T *argvars, typval_T *rettv, bool charpos)
+  if (set_chars_option(curwin, &p_fcs, false) != NULL) {
+    error = "E835: Conflicts with value of 'fillchars'";
+  } else if (set_chars_option(curwin, &p_lcs, false) != NULL) {
+    error = "E835: Conflicts with value of 'listchars'";
```
```suggestion
    error = "E834: Conflicts with value of 'listchars'";
```

zeertzjq commented on src/nvim/eval/funcs.c:
```diff
@@ -8132,6 +8132,156 @@ static void set_position(typval_T *argvars, typval_T *rettv, bool charpos)
+        break;
+      }
+      if (set_chars_option(wp, &wp->w_p_fcs, false) != NULL) {
+        error = "E834: Conflicts with value of 'fillchars'";
```
```suggestion
        error = "E835: Conflicts with value of 'fillchars'";
```

Commit:
fix: error messages

Co-authored-by: zeertzjq <zeertzjq@outlook.com>

Commit:
fix: error messages

Co-authored-by: zeertzjq <zeertzjq@outlook.com>

zeertzjq commented:
This is too hard to rebase. Restarted in #19674
